dbms/src/Interpreters/Aggregator.cpp

```diff
@@ -701,49 +784,77 @@ ALWAYS_INLINE void Aggregator::executeImplBatch(
     {
         /// For all rows.
         AggregateDataPtr place = aggregates_pool->alloc(0);
-        for (size_t i = 0; i < agg_size; ++i)
+#define HANDLE_AGG_EMPLACE_RESULT                                                                           \
+    if likely (emplace_result_hold.has_value())                                                             \
+    {                                                                                                       \
+        if constexpr (collect_hit_rate)                                                                     \
+        {                                                                                                   \
+            ++agg_process_info.hit_row_cnt;                                                                 \
+        }                                                                                                   \
+                                                                                                            \
+        if constexpr (only_lookup)                                                                          \
+        {                                                                                                   \
+            if (!emplace_result_hold.value().isFound())                                                     \
+                agg_process_info.not_found_rows.push_back(i);                                               \
+        }                                                                                                   \
+        else                                                                                                \
+        {                                                                                                   \
+            emplace_result_hold.value().setMapped(place);                                                   \
+        }                                                                                                   \
+        processed_rows = i;                                                                                 \
+    }                                                                                                       \
+    else                                                                                                    \
+    {                                                                                                       \
+        LOG_INFO(log, "HashTable resize throw ResizeException since the data is already marked for spill"); \
+        break;                                                                                              \
+    }
+
+        std::vector<size_t> hashvals;
+        std::optional<size_t> processed_rows;
+        if constexpr (enable_prefetch)
         {
-            auto emplace_result_hold = emplaceOrFindKey<only_lookup>(
-                method,
-                state,
+            getHashVals(
```

Comment:
I think `getHashVals` may be called multiple times for the same block, maybe we can save `hashvals` into `agg_process_info` to reuse the result?

Reply:
done